Source/bmalloc/libpas/src/libpas/pas_report_crash.c

```diff
@@ -97,57 +110,68 @@ kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, m
 
         addr64_t key = (addr64_t) hash_map_entry->key;
 
-        // Lower PGM Bounds Checking
-        addr64_t bottom = (addr64_t) (key - pgm_metadata->mem_to_waste - pgm_metadata->page_size);
-        addr64_t top = (addr64_t) (key - pgm_metadata->mem_to_waste);
-
-        if ((fault_address >= bottom)
-            && (fault_address < top)) {
-            report->error_type = "out-of-bounds";
-            report->confidence = "high";
-            report->fault_address = fault_address;
-            report->allocation_size = pgm_metadata->allocation_size_requested;
-            return KERN_SUCCESS;
-        }
-
-        // Upper PGM Bounds Checking
-        bottom = (addr64_t) (key - pgm_metadata->mem_to_waste + pgm_metadata->size_of_data_pages);
-        top = (addr64_t) (key - pgm_metadata->mem_to_waste + pgm_metadata->size_of_data_pages + pgm_metadata->page_size);
-
-        if ((fault_address >= bottom)
-            && (fault_address < top)) {
-            report->error_type = "out-of-bounds";
-            report->confidence = "high";
-            report->fault_address = fault_address;
-            report->allocation_size = pgm_metadata->allocation_size_requested;
-            return KERN_SUCCESS;
-        }
-
-        // UAF Check
-        bottom = (addr64_t) key;
-        top = (addr64_t) (key - pgm_metadata->mem_to_waste + pgm_metadata->size_of_data_pages);
-
-        if ((fault_address >= bottom)
-            && (fault_address < top)) {
-            report->error_type = "use-after-free";
-            report->confidence = "high";
-            report->fault_address = fault_address;
-            report->allocation_size = pgm_metadata->allocation_size_requested;
-            return KERN_SUCCESS;
-        }
-
-        // UAF + OOB Check
-        bottom = (addr64_t) (key - pgm_metadata->mem_to_waste - pgm_metadata->page_size);
-        top = (addr64_t) key;
-
-        if ((fault_address >= bottom)
-            && (fault_address < top)) {
-            report->error_type = "OOB + UAF";
-            report->confidence = "low";
-            report->fault_address = fault_address;
-            report->allocation_size = pgm_metadata->allocation_size_requested;
-            return KERN_SUCCESS;
-        }
+        if (pgm_metadata->right_align) {
+            report->alignment = "address right-aligned";
+
+            // Right-aligned "Lower PGM OOB" checking
+            addr64_t bottom = (addr64_t) (key - pgm_metadata->mem_to_waste - pgm_metadata->page_size);
+            addr64_t top = (addr64_t) (key - pgm_metadata->mem_to_waste);
+
+            if (pas_fault_address_is_in_bounds(fault_address, bottom, top))
+                pas_update_report_crash_fields(report, "out-of-bounds", "high", fault_address, pgm_metadata->allocation_size_requested);
+
+            // Right-aligned "UAF + OOB" Check towards lower guard page
+            bottom = (addr64_t) (key - pgm_metadata->mem_to_waste - pgm_metadata->page_size);
+            top = (addr64_t) key;
```

Comment:
This should check the "wasted memory" region:

```suggestion
            bottom = (addr64_t) (key - pgm_metadata->mem_to_waste);
            top = (addr64_t) key;
```